ports/raspberrypi/bindings/tca9555r/__init__.c

```diff
@@ -0,0 +1,743 @@
+/*
```

Comment:
Is there something rp2040-specific about this code? If not, it should be placed in `$(TOPDIR)/shared-bindings` and similarly for the implementation, it should go in `$(TOPDIR)/shared-module`.

Reply:
Hey @jepler. Could you clarify the difference between `shared-bindings` and `shared-module`. I've checked both folders, and I see each has folders such as `digitalio`. It looks to me like `shared-bindings` is for the micropython binding part of a module (i.e. anything dealing with `mp_obj`)? Then `shared-module` is for `common_hal` functions, which I don't explicitly have.

It could just be i'm not thinking clearly as it's late evening over here 😅

Reply:
> Could you clarify the difference between `shared-bindings` and `shared-module`.

`shared-bindings` is the layer that defines the Python API, checks any port-apnostic preconditions and converts to the `common_hal_` internal C API. `shared-modules` is for port-agnostic implementation of `common_hal_` functions.

Reply:
Okay. In that case I'm not sure how you're wanting me to move my TCA driver over to these folders then, given theirs no `comman_hal`  style functions. Is there an example you can point me to? I based my implementation on how the CYW driver for Pico W was written.

Reply:
You are doing the separation some. You just aren't designating it common-hal. For example, `tca_port_read_polarity_state` is used to parse and validate params and then call `tca_get_polarity_port`. The latter would be `common_hal_` and moved into `shared-modules`. `tca_port_read_polarity_state` would be in `shared-bindings` and documented with a Python stub comment.

`cyw43` is a bad example because it doesn't split out the limited implementation. Both `rp2pio` and `picodvi` are split along with everything else in `shared-bindings`.

Reply:
Okay, i'll have a look at `rp2pio` and `picodvi` tomorrow.

Reply:
Still figuring this out. Splitting out the python binding stuff is obvious so I will try that first. In the meantime I've added the python stub comments you mentioned.

Reply:
@jepler Is there a process for sorting out the makefiles for having your code in `shared-bindings` and `shared-module`? I'm getting so many linker errors 😭 
```
- Verbosity options: any combination of "steps commands rules", as `make V=...` or env var BUILD_VERBOSE
QSTR not updated
Module registrations not updated
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/py/objmodule.o:(.rodata.mp_builtin_module_table+0x214): undefined reference to `tca_module'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/board.o: in function `reset_board':
board.c:(.text.reset_board+0x6): undefined reference to `common_hal_tca_set_output_port'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: board.c:(.text.reset_board+0xe): undefined reference to `common_hal_tca_set_polarity_port'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: board.c:(.text.reset_board+0x16): undefined reference to `common_hal_tca_set_config_port'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: board.c:(.text.reset_board+0x1e): undefined reference to `common_hal_tca_set_output_port'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: board.c:(.text.reset_board+0x26): undefined reference to `common_hal_tca_set_polarity_port'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: board.c:(.text.reset_board+0x2e): undefined reference to `common_hal_tca_set_config_port'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_validate_pin':
DigitalInOut.c:(.text.common_hal_digitalio_validate_pin+0x2): undefined reference to `validate_obj_is_free_pin_including_tca'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_validate_pin_or_none':
DigitalInOut.c:(.text.common_hal_digitalio_validate_pin_or_none+0x2): undefined reference to `validate_obj_is_free_pin_including_tca_or_none'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_construct':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_construct+0x24): undefined reference to `common_hal_tca_gpio_get_config'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_construct+0x34): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_switch_to_input':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_switch_to_input+0x44): undefined reference to `common_hal_tca_gpio_set_config'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_switch_to_input+0x50): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_switch_to_output':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_switch_to_output+0x6c): undefined reference to `common_hal_tca_gpio_set_output'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_switch_to_output+0x76): undefined reference to `common_hal_tca_gpio_set_config'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_switch_to_output+0xac): undefined reference to `common_hal_tca_gpio_set_output'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_switch_to_output+0xb4): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_set_value':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_set_value+0x46): undefined reference to `common_hal_tca_gpio_set_output'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_set_value+0x4c): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_get_value':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_get_value+0x26): undefined reference to `common_hal_tca_gpio_get_output'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_get_value+0x2c): undefined reference to `common_hal_tca_gpio_get_input'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_get_value+0x34): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_set_drive_mode':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_set_drive_mode+0x3e): undefined reference to `common_hal_tca_gpio_get_output'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_set_drive_mode+0x6a): undefined reference to `common_hal_tca_gpio_get_input'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_set_drive_mode+0x84): undefined reference to `common_hal_tca_gpio_set_output'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_set_drive_mode+0x8c): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_set_pull':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_set_pull+0x44): undefined reference to `common_hal_tca_gpio_set_config'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_set_pull+0x50): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_get_pull':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_get_pull+0x3c): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/digitalio/DigitalInOut.o: in function `common_hal_digitalio_digitalinout_get_reg':
DigitalInOut.c:(.text.common_hal_digitalio_digitalinout_get_reg+0x24): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/microcontroller/Pin.o: in function `common_hal_reset_pin':
Pin.c:(.text.common_hal_reset_pin+0x54): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/microcontroller/Pin.o: in function `claim_pin':
Pin.c:(.text.claim_pin+0x1c): undefined reference to `tca_pin_type'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/common-hal/microcontroller/Pin.o:Pin.c:(.text.common_hal_mcu_pin_is_free+0x40): more undefined references to `tca_pin_type' follow
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xbc): undefined reference to `pin_TCA0_6'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xc4): undefined reference to `pin_TCA0_7'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xcc): undefined reference to `pin_TCA0_12'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xd4): undefined reference to `pin_TCA0_13'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xdc): undefined reference to `pin_TCA0_14'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xe4): undefined reference to `pin_TCA0_15'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xec): undefined reference to `pin_TCA0_11'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xf4): undefined reference to `pin_TCA1_1'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0xfc): undefined reference to `pin_TCA1_2'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0x104): undefined reference to `pin_TCA1_3'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0x10c): undefined reference to `pin_TCA1_4'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0x114): undefined reference to `pin_TCA1_0'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0x11c): undefined reference to `pin_TCA1_8'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_module_globals_table+0x124): undefined reference to `pin_TCA1_9'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot1_obj+0x1c): undefined reference to `pin_TCA0_3'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot1_obj+0x20): undefined reference to `pin_TCA0_4'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot1_obj+0x24): undefined reference to `pin_TCA0_5'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot2_obj+0x1c): undefined reference to `pin_TCA0_0'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot2_obj+0x20): undefined reference to `pin_TCA0_1'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot2_obj+0x24): undefined reference to `pin_TCA0_2'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot3_obj+0x1c): undefined reference to `pin_TCA0_8'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot3_obj+0x20): undefined reference to `pin_TCA0_9'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot3_obj+0x24): undefined reference to `pin_TCA0_10'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot4_obj+0x1c): undefined reference to `pin_TCA1_7'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot4_obj+0x20): undefined reference to `pin_TCA1_5'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot4_obj+0x24): undefined reference to `pin_TCA1_6'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot5_obj+0x1c): undefined reference to `pin_TCA1_15'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot5_obj+0x20): undefined reference to `pin_TCA1_14'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot5_obj+0x24): undefined reference to `pin_TCA1_13'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot6_obj+0x1c): undefined reference to `pin_TCA1_10'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot6_obj+0x20): undefined reference to `pin_TCA1_12'
/usr/lib/gcc/arm-none-eabi/10.3.1/../../../arm-none-eabi/bin/ld: build-pimoroni_yukon/boards/pimoroni_yukon/pins.o:(.rodata.board_slot6_obj+0x24): undefined reference to `pin_TCA1_11'
```